### ISSUE TITLE
Allow configuring MAXIMUM_MEMORY at load time for 2gb vs 4gb?

### ISSUE THREAD

**Issue by binji** [emscripten-core/emscripten#16197]
Currently you can pass `-s MAXIMUM_MEMORY=4gb` to emcc and it will generate JS glue code that supports addresses up to that limit. This is convenient for browsers that support 4gb, but not so convenient for sharing this same bundle with browsers that only support 2gb limit. You can pass in your own `wasmMemory` object that has a smaller limit, but `_emscripten_resize_heap()` hardcodes the maximum heap size into the generated JS code.

This means that if the `requestedSize` is very close to the 2gb limit, it's possible that `_emscripten_resize_heap()` will actually try to grow larger than 2gb and fail (where previously it would have been clamped by this code:

```
  var newSize = Math.min(maxHeapSize, alignUp(Math.max(requestedSize, overGrownHeapSize), 65536));
```

Is it possible to have `MAXIMUM_MEMORY` be configurable via a "pre-js" file to prevent this behavior?

**Comment by kripken** [comment]
Interesting question!

We could add such a hook, but in general each time we add such things they increase code size unless they are opt-in. Another code size issue is that builds supporting >2GB have larger JS (since we need to emit more `>>>` operations in various places, to keep pointers unsigned). So if we wanted this, it could be a new flag perhaps.

However, after thinking on this, I think it can be done in a not too bad way without system changes. The function `emscripten_realloc_buffer` is where we attempt to reallocate the buffer. You could link in a user JS library that renames that function and overrides it, and then decide whether to fail or not at runtime as your project wants. Something like this:
```js
// lib.js

LibraryManager.library.$orig$emscripten_realloc_buffer = LibraryManager.library.$emscripten_realloc_buffer;

LibraryManager.library.$emscripten_realloc_buffer__deps = ['$orig$emscripten_realloc_buffer'];

LibraryManager.library.$emscripten_realloc_buffer = function(size) {
  if (Math.random() < 0.75) {
    console.log('request denied');
    return 0;
  }

  console.log('allowed');
  return orig$emscripten_realloc_buffer(size);
};
```
and then build with `--js-library lib.js`, and also with `MAXIMUM_MEMORY=4GB`.

Would that be good enough for you?

**Comment by binji** [comment]
Thanks, @kripken! Yeah, I think this might work for us, I'll get back to you.

**Comment by binji** [comment]
@kripken So it looks like it will work for us, thank you! But we're a little concerned about hacking into the emscripten internals here, since it might silently change behavior when we update emscripten. Is there a way we can error out at build time if the emscripten version changes? Or maybe if the API changes?

**Comment by sbc100** [comment]
Its kind of a shame we can't discover the maximum at runtime given the `WebAssembly.Memory` object.   Its not possible to do that via the JS API is it?

**Comment by sbc100** [comment]
I have a change up that should allow the hack to be a bit less hacky: #16255

If that lands, you can simply override `emscripten_get_heap_max` rather than having to mess around with caching the existing version.

**Comment by sbc100** [comment]
> I have a change up that should allow the hack to be a bit less hacky: #16255
> 
> If that lands, you can simply override `emscripten_get_heap_max` rather than having to mess around with caching the existing version.

That will also mean that any callers of `emscripten_get_heap_max` will get the right answer.

**Comment by binji** [comment]
> Its kind of a shame we can't discover the maximum at runtime given the WebAssembly.Memory object. Its not possible to do that via the JS API is it?

I think it's possible with this proposal: https://github.com/WebAssembly/js-types/blob/main/proposals/js-types/Overview.md

> That will also mean that any callers of emscripten_get_heap_max will get the right answer.

Nice! Though it still could potentially break out from under us, right?

**Comment by sbc100** [comment]
> Nice! Though it still could potentially break out from under us, right?

Yes, but its simpler and, I think, *less* fragile.

The alternative is that we add an extra import (e.g. Module['maxMemory']) but its hard to do that without either:

a) Increasing codeside for all users by a few bytes
b) Adding yet another setting/option

**Comment by binji** [comment]
Seems like we do need @sbc100's change. I thought we would be able to use @kripken's override, but the problem is this: `emscripten_resize_heap()` first checks to make sure that it is possible to satisfy the requested size; if not it returns false immediately. Then it attempts to choose a larger size, and gradually eases off if that size is too large. The problem with overriding `emscripten_realloc_buffer()` is that we want to clamp the maximum size at some smaller size than was passed to `MAXIMUM_MEMORY`. But we can't know whether that actually satisfies the requested size, since the requested size isn't passed in to `emscripten_realloc_buffer()`!

**Comment by sbc100** [comment]
My change was released in 3.1.4.

We can probably give you an even better way to override it using and opt-in Module property if we can solve how to do that in: https://github.com/emscripten-core/emscripten/pull/16263.   

But for now it seems like you issue has a workaround?

**Comment by binji** [comment]
Yep, I think you're solution should be OK. I'll let you know when after we upgrade to 3.1.4. Thanks!
